### ISSUE TITLE
Slice tensor on a single-dimension mask

### ISSUE THREAD

**Issue by adamlerer** [pytorch/pytorch#370]
It would be nice to be able to write (like Numpy):

```
t = torch.Tensor(5,8)
u = t[t[0].eq(3)]
```

whereas now you would have to write something like

```
u = t[t[0].eq(3).nonzero().squeeze(0)]
```

It would be nice if this worked on arbitrary dimension, e.g.

```
u = t[:,t[:,0].eq(3)]
```

**Comment by fmassa** [comment]
This works in 1.0, as now indexing with byte tensors calls internally to `nonzero`:
```python
t = (torch.rand(3, 8) * 3).floor()
print(t[:, t[0].eq(1)])
```
and the last example by Adam also works out of the box in PyTorch
```python
In [27]: a = torch.tensor([[1,3],[1,4],[2,2]])

In [28]: idx = a[:,0] == 1

In [29]: a[idx]
Out[29]:
tensor([[1, 3],
        [1, 4]])

In [30]: idx = a[0] == 1

In [31]: a[:,idx]
Out[31]:
tensor([[1],
        [1],
        [2]])
```
